✨ Add remote cluster cache manager

JoelSpeed opened the pull request:
**What this PR does / why we need it:**
Add a remote cluster cache manager that owns the full lifecycle of
caches against workload clusters. Controllers that need to watch
resources in workload clusters should use this functionality.

This is the foundation for #2414 and #2577.

It does not currently check for connectivity failures to workload clusters. It only stops the workload cluster caches if there is an error retrieving the Cluster from the management cluster. If desired, I could add something that does a minimal "ping" to each workload cluster periodically, removing any caches that have connectivity issues. Or we could do that in a follow-up.

I have not updated MachineHealthCheck to use this. I could do that in this PR, or a follow-up.

I have not added anything for KubeadmControlPlane to use this. I could do that in this PR, or a follow-up (cc @detiber).

/priority important-soon

cc @vincepri @ncdc 

--
This replaces #2835. I've taken the code that @ncdc wrote, addressed some feedback from the previous PR and added a test suite for this code. I have not done anything regarding this comment https://github.com/kubernetes-sigs/cluster-api/pull/2835/files#r401783111 as I was not sure how to interpret the conversation

With the code as is, it is not actually being used anywhere. To be used, we need to create a `ClusterCacheTracker` and call `NewClusterCacheReconciler` within `main.go`, pass the `ClusterCacheTracker` into controllers that will need access to remote caches (eg MHC), and then, from within the controller, call `ClusterCacheTracker.Watch` when reconciling their object. Tracking which of the clusters have already been watched and deduplication is handled within the `ClusterCacheTracker`, so no tracking is needed within individual controllers.

The deduplication will not work properly however with a different set of predicates for each Watch, I'm not sure if that's a major problem or not, open to suggestions for rewrites if there's a good way to fix that if we think it's important

detiber commented:
> It does not currently check for connectivity failures to workload clusters.

Things definitely brings up some interesting questions... 
- Should it be the responsibility of the tracker cache to ensure that users won't be retrieving "stale" resources from the cache?
- Or should it be the responsibility of the consumer (potentially with help from a helper utility) to determine if they care about the Node info being "stale"?

vincepri commented:
/milestone v0.3.4

ncdc commented:
I could see us adding goroutine(s) that periodically attempt to do a live query against each workload cluster's apiserver, maybe?

vincepri commented:
I'd leave it as a client responsibility to decide what to do when connectivity cannot be established to a workload cluster. In some cases, we might want a controller to not error if we can't establish contact, in some other case we might want to return an error and go in backoff loop. 

Given that all use cases aren't clear yet, I'd leave it to clients to decide and bring up requirements in the future.

ncdc commented:
I think it should be this cluster cache manager's responsibility to evict the workload cache if connectivity is down. Otherwise, clients (controllers, really) will be unaware they're not getting updated data.

vincepri commented:
What if the lost of connectivity is temporary (e.g. upgrades), would we have the ability to recover?

ncdc commented:
We could code it to retry up to $conditions (n times, some timeout, etc) before evicting. After that, it would require either an event (create/update/delete) or a full (management cluster cache) resync to get it recreated.

JoelSpeed commented:
A benefit(?) if @ncdc's suggestion [above](https://github.com/kubernetes-sigs/cluster-api/pull/2880#issuecomment-611021960) is that the `Watch` would then fail and the client (controller) would then be aware that there are connectivity issues, which they can handle in their own manner

vincepri commented:
That sounds good, I do think we need to think about recovering from connectivity errors

ncdc commented:
@JoelSpeed how would the watch fail?

JoelSpeed commented:
@ncdc When calling `newClusterCache` (assuming it's been deleted by the connectivity failure), calling `cache.New` attempts [discovery of resources](https://github.com/kubernetes-sigs/controller-runtime/blob/dfb60c9e4653ef9c3196bd959333bd57c359f703/pkg/cache/cache.go#L129), when creating a `NewDiscoveryRestMapper`, from the API of the target cluster because we provide no `Mapper` through the options. If the discovery fails because the API server is still not available, that error should be propagated and returned by Watch

ncdc commented:
@JoelSpeed ah, that makes sense. Thanks for clarifying! My suggestion was in a different section of the code, though: a goroutine owned by the cluster cache manager/tracker/reconciler. The retries would be hidden to the other controllers. So a controller would have to be reconciling something and call `Watch` **after** the cache had already been evicted, after timing out during the retry period.

JoelSpeed commented:
@ncdc Ack yeah sorry, I follow, don't think I explained very well, I think we are on the same page.  I meant that *after* an eviction, as you've said, the controller would need to have some event to trigger the reconcile, and **then** Watch would fail. So the client controller wouldn't know that the remote cluster is broken until it is next reconciling and calls Watch

JoelSpeed commented:
Do we want to try and tackle the connection loss in this PR or should that be considered a follow up?

ncdc commented:
I like the idea of having PRs be as complete as possible, but will defer to @vincepri if he feels differently.

vincepri commented:
Connection loss seems an integral part of this package, +1 to what Andy said

JoelSpeed commented:
Ack, I can try having a go at the [suggestion](https://github.com/kubernetes-sigs/cluster-api/pull/2880#issuecomment-611021960) from @ncdc. Does anyone have any opinions about how long the time out should be, how long the polling interval should be, should these be user configurable, and a "safe" but lightweight way to check connectivity to the cluster (poll `/healthz` maybe)?

vincepri commented:
@JoelSpeed This is what CAPA is currently using to health check https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/9913ac8dbb96ed9bcffa6eba75b886cd4ef8c20b/pkg/cloud/services/elb/loadbalancer.go#L263-L269, we can work off of that

vincepri commented:
@JoelSpeed any updates on this PR?

JoelSpeed commented:
@vincepri Sorry, had got stuck on the health checking, was more challenging to work out what to do and how to test it than I thought it would be. I've pushed some health checking code though I'm not 100% sure about it, PTAL

ncdc commented on controllers/remote/cluster_cache.go:
```diff
@@ -0,0 +1,401 @@
+	Kind runtime.Object
+
+	// CacheOptions are used to specify options for the remote cache, such as the Scheme to use.
+	CacheOptions cache.Options
```
This isn't used currently. Will comment below where we could plumb it through.

ncdc commented on controllers/remote/cluster_cache.go:
```diff
@@ -0,0 +1,401 @@
+		m.watches[input.Cluster] = watchesForCluster
+	}
+
+	cache, err := m.getOrCreateClusterCache(ctx, input.Cluster)
```
Could plumb through input.CacheOptions here

ncdc commented on controllers/remote/cluster_cache.go:
```diff
@@ -0,0 +1,401 @@
+}
+
+// getOrCreateClusterCache returns the clusterCache for cluster, creating a new ClusterCache if needed.
+func (m *ClusterCacheTracker) getOrCreateClusterCache(ctx context.Context, cluster client.ObjectKey) (*clusterCache, error) {
```
Plumb through CacheOptions

ncdc commented on controllers/remote/cluster_cache.go:
```diff
@@ -0,0 +1,401 @@
+		return cache, nil
+	}
+
+	return m.newClusterCache(ctx, cluster)
```
Plumb through CacheOptions

ncdc commented on controllers/remote/cluster_cache.go:
```diff
@@ -0,0 +1,401 @@
+}
+
+// newClusterCache creates and starts a new clusterCache for cluster.
+func (m *ClusterCacheTracker) newClusterCache(ctx context.Context, cluster client.ObjectKey) (*clusterCache, error) {
```
Plumb through CacheOptions

ncdc commented on controllers/remote/cluster_cache.go:
```diff
@@ -0,0 +1,401 @@
+		return nil, errors.Wrap(err, "error fetching REST client config for remote cluster")
+	}
+
+	remoteCache, err := cache.New(config, cache.Options{Scheme: m.scheme})
```
Plumb through CacheOptions. Need to decide on the scheme. We currently copy the Manager's scheme into ClusterCacheTracker.scheme when creating the tracker. We can either

1. Use CacheOptions.Scheme if present; fall back to m.scheme
1. Remove m.scheme entirely and always/only use CacheOptions.Scheme

I think 1 gives more flexibility but could be confusing (maybe?). If we choose this option, let's document it well.

fabriziopandini replied:
I'm +1 for option 2, it is more explicit

ncdc commented on controllers/remote/cluster_cache.go:
```diff
@@ -0,0 +1,401 @@
+		return c, nil
+	}
+
+	config, err := RESTConfig(ctx, m.client, cluster)
```
Should we consider setting config.Timeout either to a hard-coded or user-specifiable value?

fabriziopandini replied:
We usually set user agent as well

ncdc commented on controllers/remote/cluster_cache.go:
```diff
@@ -0,0 +1,401 @@
+	delete(m.clusterCaches, cluster)
+}
+
+type ClusterCacheReconciler struct {
```
Let's move this back down so it's next to the other CCR functions. Also please add the godoc I never got around to doing 😄

JoelSpeed replied:
Oops, why did I put healthchecking in here 🤦 Fixed!

ncdc commented on controllers/remote/cluster_cache.go:
```diff
@@ -0,0 +1,401 @@
+	// Start the cache!!!
+	go remoteCache.Start(cc.stop)
+	// Start cluster healthcheck!!!
+	go m.healthCheckCluster(ctx, cluster, config, healthCheckPollInterval, healthCheckRequestTimeout, healthCheckUnhealthyThreshold, "/")
```
I don't think you can use `ctx` here as a means to stop this goroutine. `ctx` is passed in by a controller that calls `ClusterCacheTracker.Watch()`, which presumably is a pre-reconcile context.

Could we tie its lifecycle to the remote cache's `cc.stop`?

ncdc commented on controllers/remote/cluster_cache.go:
```diff
@@ -0,0 +1,401 @@
+// healthCheckCluster will poll the cluster's API at the path given and, if there are
+// `unhealthyThreshold` consecutive failures, will deem the cluster unhealthy.
+// Once the cluster is deemed unhealthy, the cluster's cache is stopped and removed.
+func (m *ClusterCacheTracker) healthCheckCluster(ctx context.Context, cluster client.ObjectKey, cfg *rest.Config, interval, requestTimeout time.Duration, unhealthyThreshold int, path string) {
```
This function has a lot of parameters. Should we consider either an input struct, or a new struct that has members for all these parameters?

JoelSpeed replied:
I've put this into a struct and added a validate function for that struct that populates the optional parts of this struct, does that seem sensible?

benmoss commented on controllers/remote/cluster_cache_test.go:
```diff
@@ -0,0 +1,744 @@
+limitations under the License.
+*/
+
+package remote
```
These tests are super hard to understand imo. I really dislike Ginkgo and wish this project would avoid it. But if we're going to use it, I would avoid using `It`s inside of helper functions. I think this code also uses nested `Context`s/`BeforeEach`s too much. It creates this situation where you have to constantly be scanning backwards in the code to understand how variables are scoped.

I think the best pattern is just to keep variables scoped to `It` functions and not relying on variable scoping/shadowing, and using helper functions to share code/reduce boilerplate. I know this is a big ask so I'm not going to make this a show-stopper or anything but this is what drives me crazy about Ginkgo.

fabriziopandini replied:
What about breaking tests in more files e.g.
- Cache and Predicates
- Reconcile
- Health cehcks

JoelSpeed replied:
I appreciate your concern and am also not entirely happy with how these turned out. I've fixed up the rest of the comments on this PR so will push that and will try and find some time to resist this and rework the tests to be a bit easier to follow. I think Fabrizio's suggestion of breaking the tests out into smaller files might make this more manageable so I'll start with that when I'm refactoring

JoelSpeed replied:
I've broken the tests into smaller files and reworked the reconciler test to use a table, does this make things clearer? Can have a go at the other tests as well if that makes sense

benmoss replied:
It's fine I'll keep my ginkgo gripes to myself 😸

benmoss commented on controllers/remote/cluster_cache.go:
```diff
@@ -0,0 +1,401 @@
+	// An error returned implies the health check has failed a sufficient number of
+	// times for the cluster to be considered unhealthy
+	if err != nil {
+		fmt.Printf("error: %v", err)
```
Did you mean to leave this in? If so it should be using `m.log`

JoelSpeed replied:
Good catch, I think this was me debugging, removed

benmoss commented on controllers/remote/suite_test.go:
```diff
@@ -0,0 +1,86 @@
+
+func TestAPIs(t *testing.T) {
+	RegisterFailHandler(Fail)
+
```
```suggestion
   
   flag.Set("logtostderr", "false")
   flag.Parse()
   klog.SetOutput(GinkgoWriter)
   
```

the magic incantation to get klog to stop logging to stdout/stderr

JoelSpeed replied:
Today I learned! Thank you, will update! Does this mean Ginkgo only outputs the log lines for the failed tests?

benmoss replied:
Yeah exactly, without this if you run it with the `ginkgo` CLI you get a bunch of controller logs even on success. `go test` does something weird where it hides stdout/stderr somehow, but with both you now get the logs only if they fail.

ncdc commented on controllers/remote/cluster_cache.go:
```diff
@@ -0,0 +1,401 @@
+	}
+
+	// First, check if the watch already exists
+	var exists bool
+	m.watchesLock.RLock()
+	exists = m.watchExists(input.Cluster, wi)
+	m.watchesLock.RUnlock()
+
+	if exists {
+		m.log.V(4).Info("Watch already exists", "namespace", input.Cluster.Namespace, "cluster", input.Cluster.Name, "kind", fmt.Sprintf("%T", input.Kind))
+		return nil
+	}
+
+	// Doesn't exist - grab the write lock
+	m.watchesLock.Lock()
+	defer m.watchesLock.Unlock()
+
+	// Need to check if another goroutine created the watch while this one was waiting for the lock
+	if m.watchExists(input.Cluster, wi) {
+		m.log.V(4).Info("Watch already exists", "namespace", input.Cluster.Namespace, "cluster", input.Cluster.Name, "kind", fmt.Sprintf("%T", input.Kind))
+		return nil
+	}
```
FWIW this is the conventional way of dealing with RW mutexes. We can switch away from the RW to a standard mutex if we think that makes more sense for now.

benmoss replied:
Yeah, I defer to others whether the RW mutex is worth the complexity

vincepri replied:
I mentioned this a while back, I'd like us to switch to a wrapper around https://golang.org/pkg/sync/#Map which is straightforward and does a lot of this under the hood. Go's sync.Map is optimized for low writes and frequent reads, which matches our use case here as well. 

w.r.t. performance vs readability, we should always opt for readability at this stage, and can optimize things later

vincepri commented:
/milestone v0.3.6

vincepri commented:
@JoelSpeed let me know when I can do another round of reviews

JoelSpeed commented:
As far as I'm aware, the outstanding work on this PR is as follows:
- https://github.com/kubernetes-sigs/cluster-api/pull/2880#discussion_r417333596 Potentially set user agent and timeout for the clients, waiting on suggestions for values for these
- https://github.com/kubernetes-sigs/cluster-api/pull/2880#discussion_r414031911 Need decision on whether to keep or drop RWMutex's

I'm not actively changing anything other than that at the moment

vincepri commented:
I don't have strong opinions on user agent (and it could be done later as a follow-up). Replied to the other comment about the mutex, that one can also be a follow-up though.

A few issues from the linter:

```
controllers/remote/cluster_cache.go:365:30: `reponsible` is a misspelling of `responsible` (misspell)
// ClusterCacheReconciler is reponsible for stopping remote cluster caches when
                             ^
controllers/remote/suite_test.go:65:10: Error return value of `flag.Set` is not checked (errcheck)
	flag.Set("logtostderr", "false")
	        ^
```

JoelSpeed commented:
Ack, will fix the linter, I can create issues for those follow ups if you like or just try and remember them and do them once this is merged?

vincepri commented:
Issues sounds good, I might have some time later this afternoon to tackle one of them (map mutex). After linter is fixed, I'll do another round and see if we can get this merged 🎉

JoelSpeed commented:
Great, I'll create the issues. Linter fixed

vincepri commented:
code looks good — I'll follow-up to clean up the locks and maps

Just needs to be squashed
/approve
@ncdc @benmoss ready to merge?

ncdc commented:
I'm still not sure about sync.Map. The go docs state "The Map type is specialized. Most code should use a plain Go map instead, with separate locking or coordination, for better type safety and to make it easier to maintain other invariants along with the map content."

vincepri commented:
Yeah that's fair, I'm proposing to have a helper struct that ensures type safety on the maps we want. We use it in the object tracker library today https://github.com/kubernetes-sigs/cluster-api/blob/158e99887e8b73148b0fc7db101bf5966d274f41/controllers/external/tracker.go#L45-L49 and it reads much better than locking, etc

vincepri commented:
/test pull-cluster-api-e2e

Seems it failed before because prow ran out of disk space

JoelSpeed commented on controllers/remote/cluster_cache_reconciler_test.go:
```diff
@@ -0,0 +1,309 @@
+				By(fmt.Sprintf("Checking cluster %q's cache is stopped", cluster.key.Name))
+				cc := cluster.cache
+				Eventually(func() bool {
+					return cc.stopped
```
You're right, this is a race, I'll look into fixing this

WRT to flakiness, I ran `ginkgo -v -untilItFails controllers/remote/...` and let it run for 50 loops without failure, so I'm reasonably confident it shouldn't cause flakes 🙂

ncdc replied:
Sweet, thanks! Once you address that, we can get this merged and then start the follow-ups to use it 😄

JoelSpeed commented:
@ncdc @vincepri I've pushed another commit that stops the tests from racing, let me know if you're happy with that and I'll squash it again

ncdc commented:
LGTM, thanks!

JoelSpeed commented:
That was quick, thanks @ncdc! Squashed

ncdc commented:
/lgtm

vincepri commented:
/test pull-cluster-api-test

ncdc commented:
/hold

ncdc commented:
Something is flaky in the new tests:

```
• Failure [0.452 seconds]
ClusterCache Reconciler suite
/home/prow/go/src/sigs.k8s.io/cluster-api/controllers/remote/cluster_cache_reconciler_test.go:42
  When running the ClusterCacheReconciler
  /home/prow/go/src/sigs.k8s.io/cluster-api/controllers/remote/cluster_cache_reconciler_test.go:43
    when clusters are deleted
    /home/prow/go/pkg/mod/github.com/onsi/ginkgo@v1.12.0/extensions/table/table.go:92
      when no clusters deleted [It]
      /home/prow/go/pkg/mod/github.com/onsi/ginkgo@v1.12.0/extensions/table/table_entry.go:43
      Failed after 0.000s.
      
      Expected
          <bool>: true
      to be false
      /home/prow/go/src/sigs.k8s.io/cluster-api/controllers/remote/cluster_cache_reconciler_test.go:199
```

JoelSpeed commented:
Ack, will investigate, I didn't run the tests too much after fixing the races so may have introduced a problem

ncdc commented:
/lgtm cancel

JoelSpeed commented:
I've spent the last 90 mins investigating this to work out where this could have flaked.

For this test to fail, something had to call `Stop` on the `clusterCache`, there's only two places that's done, in the health checker and in the reconciler.

If it was the health checker, this would imply the health check failed, but it's health checking the testenv API so that shouldn't have happened and I would expect it to have manifested before this point.

If it was the reconciler, this would imply the cluster didn't exist when we called reconcile, again, I would have thought this would have come up before, but, I've added an extra check to make sure the cluster exists in the cache before we call that.

I played around with timings for this and added a big sleep at the beginning and was unable to reproduce the flake, so I think this should fix it.

I'd suggest we run the tests a few times on CI and see if it flakes again before squash/merge

benmoss commented:
Investigating this too, this might be of help: https://gist.github.com/benmoss/a83b50cbc38d7b1c99a1a6387b43d0ad

benmoss commented on controllers/remote/cluster_cache_reconciler_test.go:
```diff
@@ -0,0 +1,331 @@
+				return k8sClient.Get(ctx, secretKey, &corev1.Secret{})
+			}, timeout).Should(Succeed())
+
+			testClusterKey := util.ObjectKey(testCluster)
```
```suggestion
			testClusterKey = util.ObjectKey(testCluster)
```

benmoss commented:
I ran the tests 20 times and they didn't fail, there's a compilation error making it fail right now though

vincepri commented:
Have you rebased everything on master? CI does that for us, just want to make sure your last commit is also rebased

JoelSpeed commented:
Oops, I messed up my commit staging, will fix up, rebase on latest and re-push

JoelSpeed commented:
/test pull-cluster-api-test
/test pull-cluster-api-capd-e2e

JoelSpeed commented:
/test pull-cluster-api-test

Just want to run this a few times to see if it flakes again in CI

vincepri commented:
LGTM

@ncdc @benmoss ?

@JoelSpeed need to squash

ncdc commented:
Squash away!

ncdc commented:
/lgtm
/hold cancel
